RecoVertex/BeamSpotProducer/plugins/BeamSpotCompatibilityChecker.cc

```diff
@@ -0,0 +1,242 @@
+    edm::Handle<reco::BeamSpot> beamSpotFromDBHandle;
+    iEvent.getByToken(dbBSfromEventToken_, beamSpotFromDBHandle);
+    spotDB = *beamSpotFromDBHandle;
+  }
+
+  double db_BSx0_ = spotDB.x0();
+  double db_BSy0_ = spotDB.y0();
+  double db_BSz0_ = spotDB.z0();
+  double db_Beamsigmaz_ = spotDB.sigmaZ();
+  double db_BeamWidthX_ = spotDB.BeamWidthX();
+  double db_BeamWidthY_ = spotDB.BeamWidthY();
+
+  double deltaX0 = evt_BSx0_ - db_BSx0_;
+  double deltaY0 = evt_BSy0_ - db_BSy0_;
+  double deltaZ0 = evt_BSz0_ - db_BSz0_;
+  double deltaSigmaX = evt_BeamWidthX_ - db_BeamWidthX_;
+  double deltaSigmaY = evt_BeamWidthY_ - db_BeamWidthY_;
+  double deltaSigmaZ = evt_Beamsigmaz_ - db_Beamsigmaz_;
+
+  if (verbose_) {
+    edm::LogPrint("BeamSpotCompatibilityChecker") << "BS from Event: \n" << spotEvent << std::endl;
+    edm::LogPrint("BeamSpotCompatibilityChecker") << "BS from DB: \n" << spotDB << std::endl;
+  }
+
+  auto significances = compareBS(spotDB, spotEvent);
+  std::vector<std::string> labels = {"x0", "y0", "z0", "sigmaX", "sigmaY", "sigmaZ"};
+
+  std::string msg = " |delta X0|=" + std::to_string(std::abs(deltaX0) * cmToum) +
+                    " um |delta Y0|=" + std::to_string(std::abs(deltaY0) * cmToum) +
+                    " um |delta Z0|=" + std::to_string(std::abs(deltaZ0) * cmToum) +
+                    " um |delta sigmaX|=" + std::to_string(std::abs(deltaSigmaX) * cmToum) +
+                    " um |delta sigmaY|=" + std::to_string(std::abs(deltaSigmaY) * cmToum) +
+                    " um |delta sigmaZ|=" + std::to_string(std::abs(deltaSigmaZ)) + "cm";
+  if (verbose_) {
+    edm::LogPrint("BeamSpotCompatibilityChecker") << msg.c_str() << std::endl;
+  }
+
+  for (unsigned int i = 0; i < 3; i++) {
+    auto sig = significances.at(i);
+    if (sig > throwingThreshold_) {
+      edm::LogError("BeamSpotCompatibilityChecker") << msg.c_str() << std::endl;
+      throw cms::Exception("BeamSpotCompatibilityChecker")
+          << "[" << __PRETTY_FUNCTION__ << "] \n DB-Event BeamSpot " << labels.at(i) << " distance sigificance " << sig
+          << ", exceeds the threshold of " << throwingThreshold_ << "!" << std::endl;
+    } else if (sig > warningThreshold_) {
+      edm::LogWarning("BeamSpotCompatibilityChecker") << msg.c_str() << std::endl;
+      edm::LogWarning("BeamSpotCompatibilityChecker")
+          << "[" << __PRETTY_FUNCTION__ << "]  \n  DB-Event BeamSpot " << labels.at(i) << " distance significance "
+          << sig << ", exceeds the threshold of " << warningThreshold_ << "!" << std::endl;
+    }
+  }
+}
+
+std::array<float, 6> BeamSpotCompatibilityChecker::compareBS(const reco::BeamSpot& BSA, const reco::BeamSpot& BSB) {
+  auto xS = bscheck::Significance(BSA.x0(), BSB.x0(), BSA.x0Error(), BSB.x0Error());
+  auto yS = bscheck::Significance(BSA.y0(), BSB.y0(), BSA.y0Error(), BSB.y0Error());
+  auto zS = bscheck::Significance(BSA.z0(), BSB.z0(), BSA.z0Error(), BSB.z0Error());
```

Comment:
do you think only one-dimensional comparisons are useful?

Reply:
> do you think only one-dimensional comparisons are useful?

Not sure what you mean by that.

Reply:
a squished beam spot rotated in two cases can give the same 1D significance but be very different in 2D

Reply:
>  a squished beam spot rotated in two cases can give the same 1D significance but be very different in 2D

I am trying to figure out how could such squished beamspot appear in realistic conditions. It's not really clear to me. In any case the main incompatibility issue (what makes track refit fail) is about the position of the centroids, so this should be enough. 
If you have a suggestion for a better metric let me know.